### DISCUSSION TITLE
Documentation/example on adding JRE support to a project?

### DISCUSSION THREAD

**Discussion by niloc132** [google/j2cl#239] [Q&A]
Previous discussion: https://github.com/google/j2cl/discussions/157#discussioncomment-3284875

In that comment, the very reasonable point is made that adding arbitrary JRE classes/members to j2cl may not make sense for everyone - indeed, Dmitrii and I have different "gwt-nio" projects with different aims (one seeks complete emulation, the other seeks to limit itself to what can be efficiently done in the browser). 

We're in the process of providing a [flatbuffers](https://github.com/google/flatbuffers) implementation for GWT, and hoping to also make it J2CL compatible, and have been trying to use closure-compiler and guava as a guide here. However, both of those use only the stock JRE classes that can be found in j2cl (naturally, being Google projects), while FlatBuffers requires `java.nio.ByteBuffer`.

It is fairly straightforward when using Java 8 to just add `java.*` classes to a project, but with Java 11 and Bazel, I don't see any examples or documentation showing how this might be done. Is this the kind of thing that is intended to be supported but isn't? I would rule out forking j2cl itself and adding support directly with no hope of upstreaming it.

We've explored a few options to build this ourselves, but before picking an approach that works for Maven and might not work for Bazel, I was hoping to see if there was an "official" answer, or if we should just target our FlatBuffers implementation as "GWT and j2cl+maven only".

**Comment by gkdn** [comment]
Like you pointed, with Java 11 providing JRE emulation externally got more problematic. The main problem is mainly around adding new classes to existing modules. 
However in this particular case since `java.nio` is not claimed by J2CL's JRE, I expect it to be pretty straightforward. There is actually a `java.nio` implementation internally used by a particular project (and again like you pointed something that works for _their_ use case). The definition is a regular `j2cl_library` target where other targets add a dependency to.

Is that not working? If not, maybe something is different in Bazel that might be causing a problem...

**Reply by niloc132** [reply]
Good to know - in that case it is likely I made a mistake, I'll try again and get back to you. My first real attempt was to plumb something simple in the helloworld sample, but it was indeed in a plain `java.*` package (since those types tend to be simpler to set up and use quickly). I only briefly attempted java.nio before switching to helloworld, and could easily have made a mistake in that setup.

That said, there are [`java.nio.charset`](https://github.com/google/j2cl/tree/master/jre/java/java/nio/charset) classes in j2cl - are you sure that wouldn't interfere?

**Reply by niloc132** [reply]
Reporting back: it does indeed work to declare `java.nio` classes in their own `j2cl_library`, in the helloworld sample - tested with just the `ReadOnlyBufferException`. Will apply this to the whole library and make a post when I have something j2cl+bazel compatible to share. 

Thanks for the quick advice, and looking forward to a more general option to add other missing types (as FlatBuffers and ByteBuffers naturally suggest some additional io classes...). I'm not marking this as "answered" for the moment, unless you anticipate that this is the answer for the forseeable future, that only which are entirely unused by j2cl's jre can have emulation provided.

**Reply by gkdn** [reply]
Yes currently I only know a solution for making additional JRE that is not sharing a package. Anything beyond that we'll need to check with Bazel Java folks to see if there is a workaround that we haven't tried before.

**Reply by niloc132** [reply]
Spitballing:
What would you think of a bazel project structure that does its best (i.e. allowing for cycles) to provide a BUILD file and j2cl_library() for each `java.*` package, so that they can be selectively replaced?

Especially in light of https://github.com/google/j2cl/issues/240, it might make sense anyway to split the jre.js file (with its `goog.provide()` calls) into it own j2cl_library() anyway, so that it can have a distinct closure_js_library()-generated `.i.js` file?

**Reply by gkdn** [reply]
The situation around default JRE is more hairy; base JRE is not simply a j2cl_library; we need to create a bootclasspath and jdk system to compile. Also I don't think JDK is separable per package due to cycles; I believe that's one of the reasons it took quite bit of work to end up in current JDK module system. We would probably end up needing a set of sources for something to work instead of "libraries". 
Also even with all of that, I'm not sure around the idea of having everything replaceable in JRE; sure will hit us back down the road :)

**Reply by treblereel** [reply]
@gkdn @niloc132  I use a tweaked version of j2cl-m-p that is able to exclude .java from compilation to bytecode (useful in case of maven reactor project has a module with such classes like nio) so I can use nio with java11.
So, in fact, I feed j2cl transpiler only with nio sources. My maven nio artifact contains only java classes and I never compiler them to the bytecode.

my setup:
[https://github.com/treblereel/j2cl/tree/v20240622-1](https://github.com/treblereel/j2cl/tree/v20240622-1)
[https://github.com/treblereel/j2clmavenplugin/tree/v20240622-1_java11](https://github.com/treblereel/j2clmavenplugin/tree/v20240622-1_java11)
[https://github.com/treblereel/gwt-nio/tree/development](https://github.com/treblereel/gwt-nio/tree/development)

Did I miss any pitfalls?

**Reply by gkdn** [reply]
I don't exactly understand how you setup and maven works but if you supply same Java sources to multiple different J2CL transpilations, when you bundle them together at the end you will have problems.

**Reply by niloc132** [reply]
Minor follow up on this - adding java.nio as a distinct package isn't sufficient for some use cases, as `java.lang.Readable` references `java.nio.CharBuffer`. This impacts projects such as protobuf-java and its `TextFormat` class:
https://github.com/protocolbuffers/protobuf/blob/158312dc8c58efc777849dc89734f55c696eae1e/java/core/src/main/java/com/google/protobuf/TextFormat.java#L2045-L2059

The current policy makes this impossible:
- can't ask j2cl to add Readable without all of java.nio
- can't add a type in java.lang as part of java.nio ("The main problem is mainly around adding new classes to existing modules.")

The only alternative seems to be that individual downstream projects need to fork j2cl's JRE to add this.

(Note that we're only idly looking at j2cl support at this time, but we have straightforward GWT support now for an openrewrite-mediated fork/subset of protobuf-java and grpc-java.)
